python/tvm/relay/graph_runtime_codegen.py

```diff
@@ -0,0 +1,394 @@
+            self.let_bind(param, node)
+
+        # Then we compile the body into a graph which can depend
+        # on input variables.
+        output_ref = self.visit(func.body)
+
+        # Finally we retreive return value of program, which will
+        # become our output node.
+        self.get_node(output_ref).is_output = True
+
+    def visit_let(self, let):
+        """Visit the Let binding, by first traversing its value,
+           then setting the metadata on the returned NodeRef.
+
+           Finally visit the body, and return the NodeRef corresponding
+           to it.
+        """
+        ident = let.var
+        val = let.value
+        body = let.body
+
+        # Need to add type info?
+        val_ref = self.visit(val)
+        dtype, shape = from_tensor(val.checked_type())
+        val_node = self.get_node(val_ref)
+        val_node.attrs["dtype"] = dtype
+        val_node.attrs["shape"] = shape
+        self.add_binding(ident, val_ref)
+        return self.visit(body)
+
+    def visit_var(self, ident):
+        return self.lookup(ident)
+
+    def visit_call(self, call):
+        """Transform a ::tvm.relay.Call into an operator in the TVM graph."""
+        inputs = []
+        for arg in call.args:
+            inputs.append(self.visit(arg).to_json())
+
+        if isinstance(call.op, Op):
+            raise Exception("Operators should be transformed away; try applying the fuse_ops transformation to the expression.")
+        elif isinstance(call.op, GlobalVar):
+            func = self.env[call.op]
+        elif isinstance(call.op, Function):
+            func = call.op
+        else:
+            raise Exception("TVM runtime does not support calls to {0}".format(type(call.op)))
+
+        if int(func.attrs.Primitive) != 1:
+            raise Exception("TVM only support calls to primitive functions (i.e functions composed of fusable operator invocations)")
+
+        op_name = func.attrs.LoweredFunc.name
+
+        attrs = {'shape': shape_to_json(call.checked_type.shape),
+                 'dtype': call.checked_type.dtype}
+        call_hash = str(ir_pass.structural_hash(call))
+        op_node = OpNode("call_" + call_hash, attrs, op_name, inputs, {})
+        return self.add_node(op_node)
+
+    def to_json(self):
+        """Convert the sequence of nodes stored by the compiler into the
+           JSON format defined in: https://docs.tvm.ai/dev/nnvm_json_spec.html.
+        """
+        nodes = []
+        # First we compute "nodes" field.
+        for node in self.nodes:
+            nodes.append(node.to_json())
+
+        arg_nodes = []
+        heads = []
+        # Compute "arg_nodes" and "heads" fields.
+        for i, node in enumerate(self.nodes):
+            if isinstance(node, InputNode):
+                arg_nodes.append(i)
+
+            if node.is_output:
+                # Need to fix this.
+                heads.append(NodeRef(i).to_json())
+
+        # Compute "node_row_ptr".
+        # TODO
```

Comment:
TODO what?

Reply:
The above comment, currently I'm using NNVM to fill the field in.